internal/helm/chart/secureloader/directory.go

```diff
@@ -0,0 +1,253 @@
+		maxSize:      maxSize,
+		rules:        rules,
+		files:        make([]*loader.BufferedFile, 0),
+	}
+}
+
+func (w *secureFileWalker) walk(name, absName string, fi os.FileInfo, err error) error {
+	n := strings.TrimPrefix(name, w.absChartPath)
+	if n == "" {
+		// No need to process top level. Avoid bug with helmignore .* matching
+		// empty names. See issue 1779.
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// Normalize to / since it will also work on Windows
+	n = filepath.ToSlash(n)
+
+	if fi.IsDir() {
+		// Directory-based ignore rules should involve skipping the entire
+		// contents of that directory.
+		if w.rules.Ignore(n, fi) {
+			return filepath.SkipDir
+		}
+		// Check after excluding ignores to provide the user with an option
+		// to opt-out from including certain paths.
+		if _, err := isSecureAbsolutePath(w.root, absName); err != nil {
+			return fmt.Errorf("cannot load '%s' directory: %w", n, err)
+		}
+		return nil
+	}
+
+	// If a .helmignore file matches, skip this file.
+	if w.rules.Ignore(n, fi) {
+		return nil
+	}
+
+	// Check after excluding ignores to provide the user with an option
+	// to opt-out from including certain paths.
+	if _, err := isSecureAbsolutePath(w.root, absName); err != nil {
+		return fmt.Errorf("cannot load '%s' file: %w", n, err)
+	}
+
+	// Irregular files include devices, sockets, and other uses of files that
+	// are not regular files. In Go they have a file mode type bit set.
+	// See https://golang.org/pkg/os/#FileMode for examples.
+	if !fi.Mode().IsRegular() {
+		return fmt.Errorf("cannot load irregular file %s as it has file mode type bits set", n)
+	}
+
+	// Confirm size it not outside boundaries
+	if fileSize := fi.Size(); w.maxSize > 0 && fileSize > w.maxSize {
+		return fmt.Errorf("cannot load file %s as file size (%d) exceeds limit (%d)", n, fileSize, w.maxSize)
+	}
+
+	data, err := os.ReadFile(absName)
+	if err != nil {
+		if pathErr := new(fs.PathError); errors.As(err, &pathErr) {
+			err = &fs.PathError{Op: pathErr.Op, Path: strings.TrimPrefix(absName, w.root), Err: pathErr.Err}
+		}
+		return fmt.Errorf("error reading %s: %w", n, err)
+	}
+	data = bytes.TrimPrefix(data, utf8bom)
+
+	w.files = append(w.files, &loader.BufferedFile{Name: n, Data: data})
+	return nil
+}
+
+// isSecureAbsolutePath attempts to make the given absolute path relative to
+// root and securely joins this with root. If the result equals absolute path,
+// it is safe to use.
+func isSecureAbsolutePath(root, absPath string) (bool, error) {
+	root, absPath = filepath.Clean(root), filepath.Clean(absPath)
+	if root == "/" {
+		return true, nil
+	}
+	unsafePath, err := filepath.Rel(root, absPath)
+	if err != nil {
+		return false, fmt.Errorf("cannot calculate path relative to root for absolute path")
+	}
+	safePath, err := securejoin.SecureJoin(root, unsafePath)
+	if err != nil {
+		return false, fmt.Errorf("cannot securely join root with resolved relative path")
+	}
+	if safePath != absPath {
```

Comment:
Does this allow a symlink inside `templates/deploy.yaml` to point to a file in a dir on the same level as templates e.g. `files/deploy.yaml`?

Reply:
This function only deals with absolute paths that went through `filepath.EvalSymlinks`, as `symwalk` does this before calling `walk`.

Reply:
So symlinks with relative path inside the chart root will work?

Reply:
Yes, they will work. 

What happens is that `sympath.Walk` will start walking over the absolute chart path, any file or symlink it then finds is resolved using eval, and then passed on to `secureFileWalker.walk`. The `walk` method will then check if the absolute path (as resolved by eval) is relative to the root. Hence, `/var/something` will not be allowed, but `/tmp/workdir/something` will.

Reply:
Thanks